src/main/java/com/coreos/jetcd/EtcdLeaseImpl.java

```diff
@@ -0,0 +1,320 @@
+    /**
+     * gRPC channel
+     */
+    private ManagedChannel channel;
+    /**
+     * Timer schedule to send keep alive request
+     */
+    private ScheduledExecutorService keepAliveSchedule;
+    private ScheduledFuture<?> scheduledFuture;
+    private long scanPeriod;
+
+    private Map<Long, Lease> keepAlives = new ConcurrentHashMap<>();
+
+    /**
+     * The first time interval
+     */
+    private long firstKeepAliveTimeOut = DEFAULT_TTL;
+
+    /**
+     * KeepAlive Request Stream, put request into this stream to keep the lease alive
+     */
+    private StreamObserver<LeaseKeepAliveRequest> keepAliveRequestStreamObserver;
+
+    /**
+     * KeepAlive Response Streamer, receive keep alive response from this stream and update the
+     * nextKeepAliveTime and deadline of the leases.
+     */
+    private StreamObserver<LeaseKeepAliveResponse> keepAliveResponseStreamObserver;
+
+    public EtcdLeaseImpl(final ManagedChannel channel) {
+        this(channel, DEFAULT_SCAN_PERIOD);
+    }
+
+    public EtcdLeaseImpl(final ManagedChannel channel, long scanPeriod) {
+
+        /*
+        Init lease stub
+         */
+        this.channel = channel;
+        this.leaseFutureStub = LeaseGrpc.newFutureStub(this.channel);
+        this.leaseStub = LeaseGrpc.newStub(this.channel);
+        this.scanPeriod = scanPeriod;
+    }
+
+    /**
+     * Init the request stream to etcd
+     * start schedule to keep heartbeat to keep alive and remove dead leases
+     *
+     * @throws IllegalStateException if the service is running already
+     */
+    public void startKeepAliveService() {
+
+        synchronized (this) {
+
+            if (isKeepAliveServiceRunning()) {
+                throw new IllegalStateException("Lease keep alive service already start");
+            }
+            keepAliveResponseStreamObserver = new StreamObserver<LeaseKeepAliveResponse>() {
+                @Override
+                public void onNext(LeaseKeepAliveResponse leaseKeepAliveResponse) {
+                    processKeepAliveRespond(leaseKeepAliveResponse);
+                    Lease lease = keepAlives.get(leaseKeepAliveResponse.getID());
+                    if (lease != null && lease.isContainHandler()) {
+                        lease.getEtcdLeaseHandler().onKeepAliveRespond(leaseKeepAliveResponse);
+                    }
+                }
+
+                @Override
+                public void onError(Throwable throwable) {
+
+                }
+
+                @Override
+                public void onCompleted() {
+                }
+            };
+
+            initRequestStream(keepAliveResponseStreamObserver);
+
+            /**
+             * Start heartbeat schedule to keep alive leases and remove dead leases
+             */
+            if (keepAliveSchedule == null) {
+                keepAliveSchedule = Executors.newSingleThreadScheduledExecutor();
+            }
+            scheduledFuture = keepAliveSchedule.scheduleAtFixedRate(
```

Comment:
Can you add some comment about how this works? For new people reading the code, here is an important place to tell them how we make sure concurrency, keep alive leases, and remove dead leases.